core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java

```diff
@@ -0,0 +1,102 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Objects;
+
+import io.grpc.Attributes;
+import io.grpc.CallOptions;
+import io.grpc.LoadBalancer2.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+// TODO(lukaszx0) make this class pkg private after removing deprecated API in PickSubchannelArgs
+class PickSubchannelArgsImpl extends PickSubchannelArgs {
+  private final CallOptions callOptions;
+  private final Metadata headers;
+  private final MethodDescriptor<?, ?> method;
+
+  /**
+   * Creates call args object for given method with its call options, metadata.
+   */
+  PickSubchannelArgsImpl(CallOptions callOptions, Metadata headers, MethodDescriptor<?, ?> method) {
+    this.callOptions = checkNotNull(callOptions, "callOptions");
+    this.headers = checkNotNull(headers, "headers");
+    this.method = checkNotNull(method, "method");
+  }
+
+  @Override
+  public Attributes getAffinity() {
+    return callOptions.getAffinity();
+  }
+
+  @Override
+  public Metadata getHeaders() {
+    return headers;
+  }
+
+  @Override
+  public CallOptions getCallOptions() {
+    return callOptions;
+  }
+
+  @Override
+  public MethodDescriptor<?, ?> getMethodDescriptor() {
+    return method;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    PickSubchannelArgsImpl that = (PickSubchannelArgsImpl) o;
+    return Objects.equal(callOptions, that.callOptions)
+        && Objects.equal(headers, that.headers)
+        && Objects.equal(method, that.method);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(callOptions, headers, method);
+  }
```

Comment:
Ewww... Um, the throwing problem sort of makes me think we shouldn't implement `toString()` in the abstract class either. Granted, we could call a method within try-catch in order to handle an exception, but it is still pretty icky.

The only reason this is an abstract class instead of interface is to allow adding new methods in the future. So in that way, I don't mind requiring toString to be implemented in each extending class.